[runSofa] Experimental : Headless recorder

ErwanDouaille opened the pull request:
HeadlessRecorder is, as its name suggest, a new gui plugin who allow users to records sofa visualisation without any windows (batch only)

⚠ Linux only ⚠

https://github.com/sofa-framework/sofa/issues/537

______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

ErwanDouaille committed:
first version of HeadlessRecorder

ErwanDouaille committed:
fix missing cmake files

ErwanDouaille committed:
fix arguments error

ErwanDouaille committed:
First commit of Readme.md

ErwanDouaille committed:
some cleaning

ErwanDouaille committed:
git merge, after some cleanings

ErwanDouaille committed:
fix some warnings, removed deprecated functions

ErwanDouaille committed:
Tired of trying to use egl, switch to Xgl

ErwanDouaille committed:
info message

ErwanDouaille committed:
remove dependency of libpng (use img from sofa), change cmake for linux only + ffmpeg

ErwanDouaille commented on cmake/Modules/FindFFMPEG.cmake:
```diff
@@ -0,0 +1,84 @@
+# Modified for Hedgewars by Stepik777
+#
+# Redistribution and use is allowed according to the terms of the New
+# BSD license.
```
Is that okay ? Can I include it ?

ErwanDouaille commented on applications/projects/runSofa/Main.cpp:
```diff
@@ -195,30 +195,40 @@ int main(int argc, char** argv)
-    .option(&messageHandler,'f',"formatting","select the message formatting to use (auto, clang, sofa, rich, test)")
-    .option(&enableInteraction, 'i', "interactive", "enable interactive mode for the GUI which includes idle and mouse events (EXPERIMENTAL)")
-    (argc,argv);
+            // alphabetical order on short name
```
I need to cleanup this part for the new argument parser

ErwanDouaille committed:
Merge from SofaFramework/master, fix gui argument

ErwanDouaille committed:
add argument, cleaning

ErwanDouaille committed:
add interacive camera for caduceus, plus some fix

ErwanDouaille commented on examples/Demos/caduceus.scn:
```diff
@@ -7,7 +7,7 @@
     <BruteForceDetection name="N2" />
     <MinProximityIntersection name="Proximity" alarmDistance="1.5" contactDistance="1" />
-
+    <InteractiveCamera position="0 30 90" lookAt="0 30 0"/>
```
I added an interactive camera for the caduceus example. By adding it, it's now possible to do some footage of the snake with HeadlessRecorder. This plugin need an interactive camera to  correctly place the camera.

ErwanDouaille committed:
Update README.md

ErwanDouaille committed:
Update README.md

ErwanDouaille committed:
Update README.md

ErwanDouaille committed:
remove a comment in cmakelist

ErwanDouaille committed:
fix compilation for win/mac

ErwanDouaille commented:
@guparan @hugtalbot @damienmarchal I think it's ready for the 17.12 release, as an experimental feature.

damienmarchal commented:
Congratulation to Erwan.

damienmarchal commented:
[ci-build][with-scene-tests]

damienmarchal commented:
[ci-build][with-scene-tests]

ErwanDouaille committed:
fix non called record stop.
Add sigterm handler with proper way for stop recording at any time.
Add an unlimited record time option named "recordUntilEndAnimate".
Refactor encode part, fix part of pts problems.

ErwanDouaille committed:
change default file name

ErwanDouaille committed:
cleaning

ErwanDouaille committed:
Merge branch 'HeadlessRecorder' of https://github.com/SofaDefrost/sofa into HeadlessRecorder

ErwanDouaille committed:
fix non-strictly-monotonic PTS warning

guparan commented:
[ci-build][with-scene-tests]

ErwanDouaille committed:
fix a bug of ratio with av

ErwanDouaille commented:
@damienmarchal @hugtalbot , stupid question. How can I write a test for a GUI ? I don't see any tests for qt/glut ... :/

ErwanDouaille committed:
add cmake message experimental, remove cout message, fix warning message

ErwanDouaille committed:
add a simple test. Check if video file > 30kb

ErwanDouaille commented:
@guparan can you enable the HeadlessRecorder plugin ? It looks like this plugin is not turned on on CI :/

guparan committed:
[CI] Enable SofaHeadlessRecorder in option builds

guparan commented:
Here you go.
It may be better to check if FFMPEG and X11 are installed before enabling the plugin.
EDIT: I'll add an `if` block to enable only on Linux, my bad.

guparan committed:
[CI] Restrict SofaHeadlessRecorder to Linux VMs

guparan committed:
[CI] FIX HeadlessRecorder name

guparan committed:
[CI] Remove ffmpeg check for HeadlessRecorder

ErwanDouaille commented:
[ci-build][with-scene-tests]

ErwanDouaille committed:
add missing lib installation

ErwanDouaille committed:
fix for test file

ErwanDouaille commented:
[ci-build][with-scene-tests]

guparan commented:
Any idea why it's still failing on Linux VMs? :-(

ErwanDouaille commented:
Failures are due to old ffmpeg libraries. They changed the api (10/2015 as I remember). 
Actually it´s working on a fresh ubuntu 17.10 but not on the old LTS. I don´t remember the ffmpeg version number. 

@bcarrez was working on centos and tried to fetch the correct version before pushing it to the ubuntu vm.

ErwanDouaille commented:
[ci-build][with-scene-tests]

ErwanDouaille commented:
Well, it's now compiling on linux since @bcarrez installed a new version of some ffmpeg libs.
The failure on the ubuntu build is due to the lack of X. @damienmarchal @guparan @hugtalbot do we need to test this "gui" ? I means, others gui are not tested and the problem would be the same as we encounter with the headless recorder if we consider testing it.

On centos, we need to install newer version of ffmpeg libs.

bcarrez commented:
@ErwanDouaille your remark makes sense. 
There are several points to be aware of:
- installing a X server on the slaves will take time
- it may destabilize the build slave (?) (disc space, unforeseen problem running a GUI server on a headless machine,...)
- and I agree: if we do it, the next step should be to test all existing GUIs

bcarrez commented:
Ahem... configuring a RECENT version of libavcodec and co. is a pain in the... back on Centos. I kindly suggest we disable this feature on centos...

ErwanDouaille committed:
simply add an example for screenshots

ErwanDouaille committed:
added the Interactive camera information

ErwanDouaille committed:
trying to disable headless recorder on centos

bcarrez commented:
With the new exclusion of SOFAGUI_HEADLESS_RECORDER on CentOS, looks like all tests pass correctly.
I think that once the dashboard has finished, this PR will be ready to merge.

ErwanDouaille commented:
[ci-build]

guparan commented on applications/sofa/gui/headlessRecorder/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+
+set(EXTRA_FILES
+    headlessRecorder_test
+)
```
Is this usefull?

ErwanDouaille replied:
From my point of view yes. By using this extra_files feature I can edit it from qtcreator and it is shown as part of my project

guparan replied:
But I can't see the `EXTRA_FILES` variable used anywhere. How does it have any impact on your project?

ErwanDouaille replied:
By example, qtcreator will only display files which are parts of the project, HEADER_FILES and SOURCE_FILES. If you want to display a file which is not a source or header in your project tree then EXTRA_FILES is usefull. If you don't want to make this file visible from the project tree that's ok. But please, consider than headlessRecorder_test in not a binary file. It's a bash script.

Be able to edit it from the IDE is a plus. Consider it like a cpp test file, not as a binary.

ErwanDouaille committed:
Remove the test

guparan committed:
[SofaGui] ADD SOFAGUI_BUILD_TESTS (default ON)

This variable is meant to differentiate GUI dependent tests from others.
This way we can disable GUI dependent tests on CI.

ErwanDouaille commented:
Dreams do come true ... It's green !!!

damienmarchal commented:
[ci-build][with-scene-tests]

damienmarchal commented:
So @guparan and @hugtalbot  do we set this one to ready ?

guparan committed:
Merge branch 'master' into HeadlessRecorder

ErwanDouaille commented:
@guparan thanks for the fix
